// node-opcua-basic-types/dist/floats.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/***
 * @module node-opcua-basic-types
 */
const _ = require("underscore");
const minFloat = -3.4 * Math.pow(10, 38);
const maxFloat = 3.4 * Math.pow(10, 38);
/**
 * return a random float value in the range of  min inclusive and  max exclusive
 * @method getRandomInt
 * @param min
 * @param max
 * @return {*}
 * @private
 */
function getRandomDouble(min, max) {
    return Math.random() * (max - min) + min;
}
function isValidFloat(value) {
    if (!_.isFinite(value)) {
        return false;
    }
    return value > minFloat && value < maxFloat;
}
exports.isValidFloat = isValidFloat;
function roundToFloat2(float) {
    if (float === 0) {
        return float;
    }
    // this method artificially rounds a float to 7 significant digit in base 10
    // Note:
    //   this is to overcome the that that Javascript doesn't  provide  single precision float values (32 bits)
    //   but only double precision float values
    // wikipedia:(http://en.wikipedia.org/wiki/Floating_point)
    //
    // * Single precision, usually used to represent the "float" type in the C language family
    //   (though this is not guaranteed). This is a binary format that occupies 32 bits (4 bytes) and its
    //   significand has a precision of 24 bits (about 7 decimal digits).
    // * Double precision, usually used to represent the "double" type in the C language family
    //   (though this is not guaranteed). This is a binary format that occupies 64 bits (8 bytes) and its
    //   significand has a precision of 53 bits (about 16 decimal digits).
    //
    const nbDigits = Math.ceil(Math.log(Math.abs(float)) / Math.log(10));
    const scale = Math.pow(10, -nbDigits + 2);
    return Math.round(float * scale) / scale;
}
exports.roundToFloat2 = roundToFloat2;
const r = new Float32Array(1);
function roundToFloat(float) {
    r[0] = float;
    const floatR = r[0];
    return floatR;
}
function randomFloat() {
    return roundToFloat(getRandomDouble(-1000, 1000));
}
exports.randomFloat = randomFloat;
function encodeFloat(value, stream) {
    stream.writeFloat(value);
}
exports.encodeFloat = encodeFloat;
function decodeFloat(stream) {
    return stream.readFloat();
}
exports.decodeFloat = decodeFloat;
function isValidDouble(value) {
    if (!_.isFinite(value)) {
        return false;
    }
    return true;
}
exports.isValidDouble = isValidDouble;
function randomDouble() {
    return getRandomDouble(-1000000, 1000000);
}
exports.randomDouble = randomDouble;
function encodeDouble(value, stream) {
    stream.writeDouble(value);
}
exports.encodeDouble = encodeDouble;
function decodeDouble(stream) {
    return stream.readDouble();
}
exports.decodeDouble = decodeDouble;
function coerceFloat(value) {
    if (value === null || value === undefined) {
        return value;
    }
    return parseFloat(value);
}
exports.coerceFloat = coerceFloat;
function coerceDouble(value) {
    if (value === null || value === undefined) {
        return value;
    }
    return parseFloat(value);
}
exports.coerceDouble = coerceDouble;
