// node-opcua-address-space/dist/src/data_access/ua_multistate_discrete.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * @module node-opcua-address-space.DataAccess
 */
const _ = require("underscore");
const node_opcua_assert_1 = require("node-opcua-assert");
const node_opcua_status_code_1 = require("node-opcua-status-code");
const node_opcua_variant_1 = require("node-opcua-variant");
const node_opcua_variant_2 = require("node-opcua-variant");
const ua_variable_1 = require("../ua_variable");
/**
 * @class UAMultiStateDiscrete
 */
class UAMultiStateDiscrete extends ua_variable_1.UAVariable {
    getValue() {
        return this.readValue().value.value;
    }
    getValueAsString() {
        const index = this.getValue();
        const arr = this.enumStrings.readValue().value.value;
        node_opcua_assert_1.assert(_.isArray(arr));
        return arr[index].text.toString();
    }
    getIndex(value) {
        const arr = this.enumStrings.readValue().value.value;
        node_opcua_assert_1.assert(_.isArray(arr));
        const index = arr.findIndex((a) => a.text === value);
        return index;
    }
    setValue(value) {
        if (typeof (value) === "string") {
            const index = this.getIndex(value);
            node_opcua_assert_1.assert(index >= 0, " invalid multi state value provided");
            return this.setValue(index);
        }
        node_opcua_assert_1.assert(_.isFinite(value));
        return this.setValueFromSource(new node_opcua_variant_2.Variant({ dataType: node_opcua_variant_1.DataType.UInt32, value }));
    }
    isValueInRange(value) {
        const arrayEnumStrings = this.enumStrings.readValue().value.value;
        // MultiStateDiscreteType
        node_opcua_assert_1.assert(value.dataType === node_opcua_variant_1.DataType.UInt32);
        if (value.value >= arrayEnumStrings.length) {
            return node_opcua_status_code_1.StatusCodes.BadOutOfRange;
        }
        return node_opcua_status_code_1.StatusCodes.Good;
    }
}
exports.UAMultiStateDiscrete = UAMultiStateDiscrete;
